### PR TITLE
test: Code Coverage: Add Edit Expense #10

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 404a8a2</samp>

The pull request adds and updates several test files for the add-edit-expense component and its related services, to improve the code coverage and test the functionality and edge cases of expense creation and editing. The pull request also refactors some code and mock data for clarity and consistency, and adds a new method to handle file upload logic in the add-edit-expense page.

### <samp>🤖 Generated by Copilot at 404a8a2</samp>

> _We're testing the `add-edit-expense` page_
> _With mock data and expectations to gauge_
> _We'll cover every scenario and case_
> _And refactor the code to make it more sage_

### <samp>🤖 Generated by Copilot at 404a8a2</samp>

*  Add and update mock data objects for testing various scenarios ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-ffa8f022a8cce95188ae7c79b4e1f34074d6c8086297a6822f542110666d2405L109-R109), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-647ac3e32fe821e793dda69a35ffee275d93c78bdd0a32d92fdcace1189bc7c0R18-R22), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-1cac010b51c1a7c2fbdf9bf224a7e267779365abe08b0f16df1ef46e41483951R1200-R1209), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-c7abedba9661487444ab115563a479af552ae5d0c5aabfd882e1226a85c6cb94R13-R22), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-21ae928ce640ad62cfada5fee7bf0bbc490300ab04a25f8309a9878e21b2f1cfR1765-R1802))
*  Add and update test cases for the removeCorporateCardExpense method to check the confirmation popup and the expense removal logic ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-1b36a2db6c6b83c6489707b9032ec98c82a71edc5d235175f2bd80a158ba71d6R697), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-1b36a2db6c6b83c6489707b9032ec98c82a71edc5d235175f2bd80a158ba71d6R737), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-1b36a2db6c6b83c6489707b9032ec98c82a71edc5d235175f2bd80a158ba71d6R762-R829))
*  Add and update test cases for the getActionSheetOptions method to check the handlers of each option for splitting, marking, and removing expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-1b36a2db6c6b83c6489707b9032ec98c82a71edc5d235175f2bd80a158ba71d6L925-R998), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-1b36a2db6c6b83c6489707b9032ec98c82a71edc5d235175f2bd80a158ba71d6L933-R1020))
*  Add and update test cases for the getPaymentModes method to check the payment modes based on the org settings and the card expense status ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L16-R22), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L226-R309))
*  Add and update test cases for the getParsedReceipt method to check the receipt parsing based on the receipt data and the exchange rate ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L774-R890))
*  Add and update test cases for the trackCreateExpense method to check the tracking service based on the expense properties and the preset values ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L853-R940), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-8fde0d56f6a9cf999839c0931458cdae9f882355edcd31b9b700f9fc9e538647R55-R56), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-8fde0d56f6a9cf999839c0931458cdae9f882355edcd31b9b700f9fc9e538647L596-R664), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57R20-R22), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57L356-R428))
*  Add and update test cases for the showSuggestedDuplicates method to check the modal dismissal and the duplicate expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L1137-R1272))
*  Add a test case for the onChangeCallback method to check the file upload callback ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57L246-R288))
*  Add and update test cases for the addAttachments method to check the file upload based on the platform and the file upload ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57L305-R404))
*  Add and update test cases for the trackEditExpense method to check the tracking service based on the expense properties and the preset values ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57L545-R617))
*  Add and update test cases for the getMarkDismissModalParams method to check the dismiss expense logic ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639R298-R323))
*  Add a test case for the setupBalanceFlag method to check the balance flag based on the account type ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639R356-R370))
*  Add and update test cases for the getReceiptCount method to check the receipt count based on the receipt data ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639L683-R766))
*  Add and update test cases for the ngOnInit method to check the report permissions based on the route params ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f819d5efa05ff0d70b82a4b7a544bf8cb8a000eaf147d4fc2be232844156469bR407-R428))
*  Add a new method to handle the file upload input change event and simplify the logic ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bR3877-R3881), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL3883-R3888))
*  Update the mock data objects for expense policy and custom input to test the scenarios when the services return null or undefined ([link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57L696-R812), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639L407-R484), [link](https://github.com/fylein/fyle-mobile-app/pull/2211/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639L866-R940))

## Clickup
https://app.clickup.com/t/85ztk93kr

## Code Coverage
`96.96% Statements 1406/1450`
`84.61% Branches 1303/1540`
`96.57% Functions 423/438`
`97.14% Lines 1363/1403`

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/fyle/add-edit-expense/add-edit-expense-6.spec.ts [hunk 1]
```diff
@@ -404,5 +404,27 @@ export function TestCases6(getTestBed) {
...
+      component.fg.controls.currencyObj.setValue({
+        amount: null,
+        currency: null,
+        orig_amount: 10,
+        orig_currency: 'USD',
+      });
+      fixture.detectChanges();
+
+      const result = component.currencyObjValidator(component.fg.controls.currencyObj);
+      expect(result).toBeNull();
```
Comment (Julias0) [inline src/app/fyle/add-edit-expense/add-edit-expense-6.spec.ts:427]: Can you check the inverse case too where it returns `{ required: false }` ?

**Review by Julias0 — CHANGES_REQUESTED** [review]
Will review after Dimple / Omkar

**Review by OmkarJ13 — COMMENTED** [review]
Added comments and questions

File: src/app/core/mock-data/unflattened-txn.data.ts [hunk 2]
```diff
@@ -1762,3 +1762,41 @@ export const unflattenedExpWithCCCExpn: UnflattenedTransaction = {
     },
   ],
 };
+
+export const trackCreateExpData = {
```
Comment (Dimple16) [inline src/app/core/mock-data/unflattened-txn.data.ts:1766]: aren't we adding the type?

File: src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts [hunk 3]
```diff
@@ -757,6 +759,74 @@ export function TestCases1(getTestBed) {
           ToastContent: 'Successfully removed the card details from the expense.',
         });
       }));
+
+      it('should show popup but take no action', fakeAsync(() => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts:763]: please describe which pop-up and method name if any

File: src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts [hunk 4]
```diff
@@ -757,6 +759,74 @@ export function TestCases1(getTestBed) {
...
+        expect(transactionService.getRemoveCardExpenseDialogBody).toHaveBeenCalledTimes(1);
+        expect(component.getRemoveCCCExpModalParams).toHaveBeenCalledOnceWith(header, body, ctaText, ctaLoadingText);
+        expect(popoverController.create).toHaveBeenCalledOnceWith(
+          component.getRemoveCCCExpModalParams(header, body, ctaText, ctaLoadingText)
+        );
+        expect(trackingService.unlinkCorporateCardExpense).not.toHaveBeenCalled();
+        expect(component.showSnackBarToast).not.toHaveBeenCalled();
+      }));
+
+      it('should do back to expenses page if no expense is found', fakeAsync(() => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts:792]: `should go back`

File: src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts [hunk 5]
```diff
@@ -757,6 +759,74 @@ export function TestCases1(getTestBed) {
...
+        expect(transactionService.getRemoveCardExpenseDialogBody).toHaveBeenCalledTimes(1);
+        expect(component.getRemoveCCCExpModalParams).toHaveBeenCalledOnceWith(header, body, ctaText, ctaLoadingText);
+        expect(popoverController.create).toHaveBeenCalledOnceWith(
+          component.getRemoveCCCExpModalParams(header, body, ctaText, ctaLoadingText)
+        );
+        expect(trackingService.unlinkCorporateCardExpense).toHaveBeenCalledOnceWith({
+          Type: 'unlink corporate card expense',
+          transaction: undefined,
+        });
+        expect(component.showSnackBarToast).toHaveBeenCalledOnceWith(
```
[the remark this answers is not in the document]
Comment (Dimple16) [reply src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts:821]: didn't understand why we are tracking undefined transactions, @jayfyle can you verify once?

File: src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts [hunk 6]
```diff
@@ -223,30 +228,84 @@ export function TestCases2(getTestBed) {
...
-        expect(component.showCardTransaction).toBeFalse();
-        expect(accountsService.getEMyAccounts).toHaveBeenCalledTimes(1);
-        expect(orgSettingsService.get).toHaveBeenCalledTimes(1);
-        expect(orgUserSettingsService.getAllowedPaymentModes).toHaveBeenCalledTimes(1);
-        expect(paymentModesService.checkIfPaymentModeConfigurationsIsEnabled).toHaveBeenCalledTimes(1);
-        done();
+      it('should get payment modes in case org settings are not present', (done) => {
+        component.etxn$ = of({
+          ...unflattenExp1,
+          tx: { ...unflattenExp1.tx, corporate_credit_card_expense_group_id: false },
```
[the remark this answers is not in the document]
Comment (Dimple16) [reply src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts:262]: +1
or check if it's an empty string in this case

File: src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts [hunk 7]
```diff
@@ -223,30 +228,84 @@ export function TestCases2(getTestBed) {
...
+          expect(orgUserSettingsService.getAllowedPaymentModes).toHaveBeenCalledTimes(1);
+          expect(paymentModesService.checkIfPaymentModeConfigurationsIsEnabled).toHaveBeenCalledTimes(1);
+          done();
+        });
+      });
+
+      it('should get payment modes if CCC expense is enabled', (done) => {
+        component.etxn$ = of({
+          ...unflattenExp1,
+          tx: { ...unflattenExp1.tx, corporate_credit_card_expense_group_id: false },
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts:288]: `corporate_credit_card_expense_group_id` shouldn't be a boolean

File: src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts [hunk 8]
```diff
@@ -771,35 +830,63 @@ export function TestCases2(getTestBed) {
...
-            category: 'SYSTEM',
-            currency: 'USD',
-            amount: 100,
-            date: new Date('2023-02-15T06:30:00.000Z'),
-            invoice_dt: new Date('2023-02-24T12:03:57.680Z'),
-            vendor_name: 'vendor',
-          },
-          exchangeRate: 82,
+        result.then((res) => {
+          expect(res).toEqual({
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts:843]: move this data to mock data, why add it here directly?

File: src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts [hunk 9]
```diff
@@ -771,35 +830,63 @@ export function TestCases2(getTestBed) {
...
+      it('should get parsed receipt without date', fakeAsync(() => {
+        transactionOutboxService.parseReceipt.and.resolveTo(parsedReceiptData2);
+        currencyService.getHomeCurrency.and.returnValue(of('INR'));
+        currencyService.getExchangeRate.and.returnValue(of(82));
+
+        const result = component.getParsedReceipt('base64encoded', 'jpeg');
+        tick(500);
+
+        result.then((res) => {
+          expect(res).toEqual({
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts:873]: same comment as above

File: src/app/fyle/add-edit-expense/add-edit-expense-3.spec.ts [hunk 10]
```diff
@@ -593,24 +595,73 @@ export function TestCases3(getTestBed) {
...
+          Time_Spent: '30 secs',
+          Used_Autofilled_Category: false,
+          Used_Autofilled_Project: true,
+          Used_Autofilled_CostCenter: true,
+          Used_Autofilled_Currency: true,
+          Instafyle: true,
+        });
+      });
+
+      it('should track create expense event for an expense without currency', () => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-3.spec.ts:622]: can the user add an expense without currency if manually adding and expense from the form?

File: src/app/fyle/add-edit-expense/add-edit-expense-3.spec.ts [hunk 11]
```diff
@@ -593,24 +595,73 @@ export function TestCases3(getTestBed) {
...
+          Used_Autofilled_Category: false,
+          Used_Autofilled_Project: true,
+          Used_Autofilled_CostCenter: true,
+          Used_Autofilled_Currency: true,
+          Instafyle: true,
+        });
+      });
+    });
+
+    it('should track create expense event', () => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-3.spec.ts:646]: both the tests have the same description, please add them accordingly

File: src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts [hunk 12]
```diff
@@ -693,27 +765,51 @@ export function TestCases4(getTestBed) {
...
+          Time_Spent: '300 secs',
+          Used_Autofilled_Category: true,
+          Used_Autofilled_Project: true,
+          Used_Autofilled_CostCenter: true,
+          Used_Autofilled_Currency: true,
+        });
+        expect(component.getTimeSpentOnPage).toHaveBeenCalledTimes(1);
+      });
+
+      it('should track edit expense event for an expense without currency', () => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts:792]: this is not ideally possible, right? can you check if this even happens in UI?

**Review by Dimple16 — CHANGES_REQUESTED** [review]
added comments

File: src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-1.spec.ts:821]: Here the txn is null, causing the undefined property. I have added proper data now.

File: src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts [hunk 13]
```diff
@@ -223,30 +228,84 @@ export function TestCases2(getTestBed) {
...
 
-      component.getPaymentModes().subscribe((res) => {
-        expect(res).toEqual(paymentModesData);
-        expect(component.showCardTransaction).toBeFalse();
-        expect(accountsService.getEMyAccounts).toHaveBeenCalledTimes(1);
-        expect(orgSettingsService.get).toHaveBeenCalledTimes(1);
-        expect(orgUserSettingsService.getAllowedPaymentModes).toHaveBeenCalledTimes(1);
-        expect(paymentModesService.checkIfPaymentModeConfigurationsIsEnabled).toHaveBeenCalledTimes(1);
-        done();
+      it('should get payment modes in case org settings are not present', (done) => {
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-2.spec.ts:259]: I am not sure about UI, but this case seems to be in place if the call by org settings service fails and cannot fetch the current user's settings.

File: src/app/fyle/add-edit-expense/add-edit-expense-3.spec.ts (hunk 10 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-3.spec.ts:622]: This case is added when the expense only has the `orig_currency` property and the `currency` property is missing

File: src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts [hunk 14]
```diff
@@ -243,7 +246,46 @@ export function TestCases4(getTestBed) {
...
+        fixture.detectChanges();
+
+        const dummyNativeElement = document.createElement('input');
+
+        component.fileUpload = {
+          nativeElement: dummyNativeElement,
+        };
+
+        const nativeElement = component.fileUpload.nativeElement as HTMLInputElement;
+        spyOn(nativeElement, 'click').and.callThrough();
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts:276]: I had added this from reference from some other test suite. It is no longer needed.

File: src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts (hunk 12 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts:792]: Rephrased the test description

File: src/app/core/mock-data/parsed-receipt.data.ts [hunk 15]
```diff
@@ -10,6 +10,16 @@ export const parsedReceiptData1: ParsedReceipt = {
     vendor_name: 'vendor',
   },
 };
+export const parsedReceiptData2: ParsedReceipt = {
+  data: {
+    category: 'SYSTEM',
+    currency: 'USD',
+    amount: 100,
+    date: null,
+    invoice_dt: new Date('2023-02-24T12:03:57.680Z'),
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/mock-data/parsed-receipt.data.ts:19]: Yes this does not return data, since I was reusing data I let it be. I have verified it on local, parsing return dates in string which are later converted to date objects when adding to the txn objects. I have segregated the data accordingly.

File: src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts [hunk 16]
```diff
@@ -243,7 +246,45 @@ export function TestCases4(getTestBed) {
       }));
     });
 
+    it('onChangeCallback(): should call upload file callback', fakeAsync(() => {
+      spyOn(component, 'uploadFileCallback');
+
+      const mockFile = new File(['file contents'], 'test.png', { type: 'image/png' });
+      const mockNativeElement = {
+        files: [mockFile],
+      } as unknown as HTMLInputElement;
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts:255]: why is this `as unknown as HTMLInputElement;`?

File: src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts [hunk 17]
```diff
@@ -302,33 +343,63 @@ export function TestCases4(getTestBed) {
...
+        component.presetCostCenterId = trackAddExpenseWoCurrency.tx.cost_center_id;
+        component.presetCurrency = trackAddExpenseWoCurrency.tx.orig_currency;
+        fixture.detectChanges();
+
+        component.trackAddExpense();
+        tick(500);
+        expect(component.getCustomFields).toHaveBeenCalledOnceWith();
+        expect(component.generateEtxnFromFg).toHaveBeenCalledOnceWith(component.etxn$, jasmine.any(Observable));
+        expect(trackingService.createExpense).toHaveBeenCalledOnceWith({
+          Type: 'Receipt',
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts:390]: should we move this to `expenseTrackerPropertiesData`? let's add the mock data there

**Review by Dimple16 — APPROVED** [review]
couple of comments, rest lgtm, please check them before next review

**Review by OmkarJ13 — APPROVED** [review]

File: src/app/core/mock-data/expense-policy.data.ts [hunk 18]
```diff
@@ -15,6 +15,10 @@ export const expensePolicyData: ExpensePolicy = {
   },
 };
 
+export const expensePolicyDataWoData: ExpensePolicy = {
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/mock-data/expense-policy.data.ts:18]: Just to keep track of it and linting issues

File: src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts (hunk 16 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-4.spec.ts:255]: I am mocking an input element here but it does not have all properties defined, so to avoid type errors I am using unknown and then asserting the type
